packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -306,33 +307,40 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
     setFocusedState(isFocused);
   };
 
-  let selectionManagerProxy = useMemo(() => {
-    let proxy = new Proxy(selectionManager, {
-      get(target, prop, receiver) {
-        // If the menu is closed, don't update the selected key.
-        if (prop === 'replaceSelection' && !triggerState.isOpen) {
-          return () => {};
-        }
-        return Reflect.get(target, prop, receiver);
+  useEffect(() => {
+    // Track the last filtered collection while the menu is open so that we can freeze when the
+    // user clicks outside the popover to close the menu. Prevents the menu contents from updating as the menu closes.
+    if (triggerState.isOpen) {
+      frozenCollection.current = filteredCollection;
+    }
+  }, [triggerState.isOpen, filteredCollection]);
+
+  let displayedCollection = useMemo(() => {
+    if (triggerState.isOpen) {
+      if (showAllItems) {
+        return originalCollection;
+      } else {
+        return filteredCollection;
       }
-    });
+    } else {
+      return frozenCollection.current;
```

Comment:
technically we can't read from the ref here, it's during render
https://beta.reactjs.org/reference/react/useRef Caveats

I don't know what the best alternative is yet.
Possibly, we could keep frozenCollection as state and update it when we update the current state in the `setInputValue` function

```
  let [inputValue, setInputValue] = useControlledState(
    props.inputValue,
    props.defaultInputValue ?? '',
    props.onInputChange
  );
```
->
```
  let [inputValue, _setInputValue] = useControlledState(
    props.inputValue,
    props.defaultInputValue ?? '',
    props.onInputChange
  );
  let setInputValue = useCallback((val) => {
    _setInputValue(val);
   setFrozenCollection(...
  }, [...]);
```

Reply:
urgh, forgot about how we can't read from refs in render either, poking around for an alternative. Updating the collection as a result of `setInputValue` wouldn't be sufficient I believe since the collection could change as a result of the user updating their controlled `items` after a delay (e.g. async loading)